packages/nouns-contracts/test/foundry/NounsDAOLogicV3/UpgradeToDAOV3.sol

```diff
@@ -0,0 +1,292 @@
+    }
+
+    function test_timelockV2IsUpgradable() public {
+        address timelockV2Impl = upgradeToV3();
+
+        bytes32 slot = bytes32(uint256(keccak256('eip1967.proxy.implementation')) - 1);
+        address implementation = address(uint160(uint256(vm.load(address(daoProxy.timelock()), slot))));
+        assertEq(implementation, timelockV2Impl);
+
+        targets = [address(daoProxy.timelock())];
+        values = [0];
+        signatures = ['upgradeTo(address)'];
+        address newTimelock = address(new NewTimelockMock());
+        calldatas = [abi.encode(newTimelock)];
+        vm.prank(proposer);
+        uint256 proposalId = daoProxy.propose(targets, values, signatures, calldatas, 'upgrade to 1234');
+
+        rollAndCastVote(proposer, proposalId, 1);
+        queueAndExecute(proposalId);
+
+        implementation = address(uint160(uint256(vm.load(address(daoProxy.timelock()), slot))));
+        assertEq(implementation, address(newTimelock));
+        assertEq(NewTimelockMock(payable(address(daoProxy.timelock()))).banner(), 'NewTimelockMock');
+    }
+
+    function upgradeToV3() internal returns (address) {
+        (NounsDAOExecutorV2 timelockV2, address timelockV2Impl) = deployAndInitTimelockV2();
+        uint256 proposalId = proposeToUpgradeToDAOV3(
+            address(new NounsDAOLogicV3()),
+            address(timelockV2),
+            address(daoProxy.timelock()),
+            500 ether
+        );
+        rollAndCastVote(proposer, proposalId, 1);
+        queueAndExecute(proposalId);
+
+        return timelockV2Impl;
+    }
+
+    function queueAndExecute(uint256 proposalId) internal {
+        vm.roll(block.number + daoProxy.votingPeriod() + 1);
+        daoProxy.queue(proposalId);
+
+        vm.warp(block.timestamp + daoProxy.timelock().delay());
+        daoProxy.execute(proposalId);
+    }
+
+    function rollAndCastVote(
+        address voter,
+        uint256 proposalId,
+        uint8 support
+    ) internal {
+        vm.roll(block.number + daoProxy.votingDelay() + 1);
+        vm.prank(voter);
+        daoProxy.castVote(proposalId, support);
+    }
+
+    function proposeToSendETH(
+        address proposer_,
+        address to,
+        uint256 amount
+    ) internal returns (uint256 proposalId) {
+        targets = [to];
+        values = [amount];
+        signatures = [''];
+        calldatas = [bytes('')];
+
+        vm.prank(proposer_);
+        proposalId = daoProxy.propose(targets, values, signatures, calldatas, 'send eth');
+    }
+
+    function deployAndInitTimelockV2() internal returns (NounsDAOExecutorV2 timelockV2, address timelockV2Impl) {
+        timelockV2Impl = address(new NounsDAOExecutorV2());
+        timelockV2 = NounsDAOExecutorV2(payable(address(new ERC1967Proxy(timelockV2Impl, ''))));
+        timelockV2.initialize(address(daoProxy), timelockV1.delay());
+        return (timelockV2, timelockV2Impl);
+    }
+
+    function proposeToUpgradeToDAOV3(
```

Comment:
curious if you're planning in a later commit to expand this proposal to include things like DAO admin change and sending ETH and stETH?

wondering where do you imagine us fully simulating the upgrade proposal(s)

Reply:
@eladmallel yes this would be the place to do it.
just adding it a piece at a time